### PR TITLE
Fix uppercase and lowercase check to be unicode instead of ascii

### LINKED ISSUE #23562: Some German Capital letters are not counted as Capital
### Describe the bug
If a user is using German (and Cyrillic, probably others too) keyboard for example, and enters a Capital letter for a password the validation 'o.Password.RequireUppercase = true;' does not pass.

### To Reproduce
1. Try to set a password with just one capital letter of the following list (some of the German ones only):
Ö, Ä, Ü

### Exceptions (if any)
"PasswordRequiresUpper"
"Passwords must have at least one uppercase ('A'-'Z')."

### Further technical details
- .net core 3.1
- Тhe output:
.NET Core SDK (reflecting any global.json):
Version: 3.1.100
Commit: cd82f021f4

Runtime Environment:
OS Name: Windows
OS Version: 10.0.18362
OS Platform: Windows
RID: win10-x64
Base Path: C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
Version: 3.1.0
Commit: 65f04fb6db

.NET Core SDKs installed:
1.1.10 [C:\Program Files\dotnet\sdk]
2.1.202 [C:\Program Files\dotnet\sdk]
2.1.301 [C:\Program Files\dotnet\sdk]
2.1.402 [C:\Program Files\dotnet\sdk]
2.1.403 [C:\Program Files\dotnet\sdk]
2.2.109 [C:\Program Files\dotnet\sdk]
3.0.100 [C:\Program Files\dotnet\sdk]
3.1.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.NETCore.App 1.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 1.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
- The IDE VS 2019

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #23562

### CODE REVIEW

**Comment by Kahbazi** [comment]
Does this message also need to be updated?
https://github.com/dotnet/aspnetcore/blob/c7a2c0648a0fad7dc1f98c18702d1a4c1d72bcc2/src/Identity/Extensions.Core/src/Resources.resx#L193

**Comment by Youssef1313** [comment]
@Kahbazi Yes. Thanks. I'll update it.

**Comment by Kahbazi** [comment]
I think this test needs to be updated too.
https://github.com/dotnet/aspnetcore/blob/c95ee2b051814b787b07f55ff224d03d550aafeb/src/Identity/test/Identity.Test/PasswordValidatorTest.cs#L149

**Comment by Youssef1313** [comment]
@Kahbazi Thanks again for review. I've updated the error message in the test, and also added a test case with the new behavior.

**Comment by Youssef1313** [comment]
Any idea if there is any specific file that should be updated in this PR to track the breaking change?

**Comment by Kahbazi** [comment]
I also find this type that need to be updated.
https://github.com/dotnet/aspnetcore/blob/c95ee2b051814b787b07f55ff224d03d550aafeb/src/Identity/src/Core/PasswordOptions.cs#L28

**Comment by HaoK** [comment]
I think its probably best to take a compromise approach, seems fine to change the default implementation and comments to use Unicode Upper/Lower, but I think there's no harm in leaving these virtual, and let people continue to override the behavior to implement Upper/Lower as they see fit.

**Review by HaoK — CHANGES_REQUESTED** [review]
Revert the virtual changes and I think this looks reasonable

**Comment by Youssef1313** [comment]
@HaoK, I've added virtual back. Do you want me to use `Rune` instead of char.IsUpper and char.IsLower?

**Comment by HaoK** [comment]
Yeah if @GrabYourPitchforks suggests using Rune, I think we should listen to his guidance on correctness there :)

File: src/Identity/Extensions.Core/src/PasswordValidator.cs [hunk 1]
```diff
@@ -89,33 +90,43 @@ public virtual bool IsDigit(char c)
...
         public virtual bool IsLower(char c)
         {
-            return c >= 'a' && c <= 'z';
+            string s = c.ToString();
+            foreach (Rune r in theString.EnumerateRunes())
+            {
+                if (Rune.IsLower(r)) return true;
+            }
+            return false;
         }
```
Comment (Youssef1313) [inline src/Identity/Extensions.Core/src/PasswordValidator.cs:105]: I'm unsure if I implemented it correctly. Mentioning @GrabYourPitchforks for review.

File: src/Identity/Extensions.Core/src/PasswordValidator.cs [hunk 2]
```diff
@@ -89,33 +90,43 @@ public virtual bool IsDigit(char c)
...
         public virtual bool IsLower(char c)
         {
-            return c >= 'a' && c <= 'z';
+            string s = c.ToString();
+            foreach (Rune r in s.EnumerateRunes())
+            {
+                if (Rune.IsLower(r)) return true;
+            }
+            return false;
         }
```
Comment (Youssef1313) [inline src/Identity/Extensions.Core/src/PasswordValidator.cs:105]: I'm unsure if I implemented it correctly. Mentioning @GrabYourPitchforks for review.

**Comment by GrabYourPitchforks** [comment]
The suggestion was that the methods `IsLower(char)` and `IsUpper(char)` weren't correct in the first place and that we should no longer use these methods. The issue https://github.com/dotnet/aspnetcore/issues/23562 still needs to be fleshed out, including exactly what problem it's trying to solve. The issue as reported concerns itself only with German characters, but is it really indicating a larger issue that also affects our Chinese and Japanese and Korean users? That really deserves a more thorough discussion. I think this PR, while well-intentioned, is premature until a plan of action has been settled on there.

**Comment by blowdart** [comment]
Thank you for this, but it's too big a beaking risk for now.

**Pull request closed without merging**